File: scripts/indexing/utils/filter-verbs.ts
import { spawn } from "child_process";
import path from "path";

/**
 * Filter out verbs from text using Python spacy (which is much better than the JS version)
 * @param {string} text - The text to filter
 * @param {string} language - Language code ('en', 'es', 'pt')
 * @returns {Promise<string[]>} - Filtered words without verbs
 */
export async function filterVerbsPython(
  text: string,
  language: string
): Promise<string[]> {
  return new Promise((resolve, reject) => {
    // Path to the Python script
    const scriptPath = path.join(__dirname, "filter_verbs.py");

    // Spawn Python process
    const python = spawn("python3", [scriptPath, text, language]);

    let result = "";
    let error = "";

    // Collect stdout data
    python.stdout.on("data", (data) => {
      result += data.toString();
    });

    // Collect stderr data
    python.stderr.on("data", (data) => {
      error += data.toString();
    });

    // Handle process completion
    python.on("close", (code) => {
      if (code === 0) {
        try {
          const filteredWords = JSON.parse(result);
          resolve(filteredWords);
        } catch (parseError) {
          reject(new Error(`Failed to parse Python output: ${parseError}`));
        }
      } else {
        reject(new Error(`Python process failed with code ${code}: ${error}`));
      }
    });

    // Handle process errors
    python.on("error", (err) => {
      reject(new Error(`Failed to start Python process: ${err.message}`));
    });
  });
}

/**
 * Fallback function if Python is not available
 * @param {string[]} words - Array of words to filter
 * @param {string} language - Language code
 * @returns {string[]} - Filtered words using simple rules
 */
export function filterVerbsFallback(
  words: string[],
  language: string
): string[] {
  // Simple verb pattern matching as fallback
  const verbPatterns = {
    en: [/ing$/, /ed$/, /s$/, /es$/, /er$/, /est$/],
    es: [/ar$/, /er$/, /ir$/, /ando$/, /iendo$/, /ado$/, /ido$/],
    pt: [/ar$/, /er$/, /ir$/, /ando$/, /endo$/, /ado$/, /ido$/],
  };

  const patterns =
    verbPatterns[language as keyof typeof verbPatterns] || verbPatterns.en;

  return words.filter(
    (word) => !patterns.some((pattern) => pattern.test(word))
  );
}

File: scripts/indexing/utils/filter_verbs.py
#!/usr/bin/env python3
import sys
import json
import spacy

def filter_verbs(text, language):
    """
    Filter out verbs from text, keeping only nouns, adjectives, and proper nouns.
    
    Args:
        text (str): The text to process
        language (str): Language code ('en', 'es', 'pt')
    
    Returns:
        list: Filtered words without verbs
    """
    try:
        # Map language codes to spacy models
        models = {
            'en': 'en_core_web_sm',
            'es': 'es_core_news_sm', 
            'pt': 'pt_core_news_sm'
        }
        
        # Default to English if language not supported
        model_name = models.get(language, 'en_core_web_sm')
        
        # Load the appropriate model
        nlp = spacy.load(model_name)
        
        # Process the text
        doc = nlp(text)
        
        # Keep only nouns, adjectives, and proper nouns
        filtered_words = []
        for token in doc:
            if token.pos_ in ["NOUN", "PROPN", "ADJ"]:
                # Convert to lowercase and add to list
                filtered_words.append(token.text.lower())
        
        return filtered_words
        
    except Exception as e:
        # If spacy fails, return original words as fallback
        print(f"Error processing with spacy: {e}", file=sys.stderr)
        # Simple fallback: split text into words
        words = text.lower().split()
        return [word for word in words if len(word) > 2]

def main():
    if len(sys.argv) != 3:
        print("Usage: python filter_verbs.py <text> <language>", file=sys.stderr)
        sys.exit(1)
    
    text = sys.argv[1]
    language = sys.argv[2]
    
    try:
        filtered_words = filter_verbs(text, language)
        # Output as JSON for easy parsing in Node.js
        print(json.dumps(filtered_words))
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main() 